test_conformance/extensions/cl_khr_command_buffer/basic_command_buffer.cpp

```diff
@@ -0,0 +1,532 @@
+
+        program =
+            clCreateProgramWithSource(context, 1, &kernel_str, lengths, &error);
+        test_error(error, "Failed to create program with source");
+
+        error = clBuildProgram(program, 1, &device, nullptr, nullptr, nullptr);
+        test_error(error, "Failed to build program");
+
+        in_mem = clCreateBuffer(context, CL_MEM_READ_ONLY,
+                                sizeof(cl_int) * num_elements, nullptr, &error);
+        test_error(error, "clCreateBuffer failed");
+
+        out_mem =
+            clCreateBuffer(context, CL_MEM_WRITE_ONLY,
+                           sizeof(cl_int) * num_elements, nullptr, &error);
+        test_error(error, "clCreateBuffer failed");
+
+        kernel = clCreateKernel(program, "copy", &error);
+        test_error(error, "Failed to create copy kernel");
+
+        error = clSetKernelArg(kernel, 0, sizeof(in_mem), &in_mem);
+        test_error(error, "clSetKernelArg failed");
+
+        error = clSetKernelArg(kernel, 1, sizeof(out_mem), &out_mem);
+        test_error(error, "clSetKernelArg failed");
+
+        if (simultaneous_use)
+        {
+            cl_command_buffer_properties_khr properties[3] = {
+                CL_COMMAND_BUFFER_PROPERTIES_ARRAY_KHR,
+                CL_COMMAND_BUFFER_SIMULTANEOUS_USE_KHR, 0
+            };
+            command_buffer =
+                clCreateCommandBufferKHR(1, &queue, properties, &error);
+        }
+        else
+        {
+            command_buffer =
+                clCreateCommandBufferKHR(1, &queue, nullptr, &error);
+        }
+        test_error(error, "clCreateCommandBufferKHR failed");
+
+        return CL_SUCCESS;
+    }
+
+    /*
+     * Subtests returning an OpenCL error code
+     */
+
+    // Test enqueuing a command-buffer containing a single NDRange command
+    // once
+    cl_int BasicEnqueue();
+
+    // Test enqueuing a command-buffer containing multiple command, including
+    // operations other than NDRange kernel execution.
+    cl_int MixedCommandsTest();
+
+    // Test enqueueing a command-buffer blocked on a user-event
+    cl_int UserEventTest();
+
+    // Test flushing the command-queue between command-buffer enqueues
+    cl_int ExplicitFlushTest();
+
+    // Test enqueueing a command-buffer twice separated by another enqueue
+    // operation
+    cl_int InterleavedEnqueueTest();
+
+    // Device supports the simultaneous-use capability
+    bool simultaneous_use;
+
+protected:
+    size_t data_size() const { return num_elements * sizeof(cl_int); }
+
+    cl_context context;
+    cl_command_queue queue;
+    cl_command_buffer_khr command_buffer;
+    clProgramWrapper program;
+    clKernelWrapper kernel;
+    clMemWrapper in_mem, out_mem;
+    size_t num_elements;
+};
+
+cl_int BasicCommandBufferTest::BasicEnqueue()
+{
+    cl_int error = clCommandNDRangeKernelKHR(
+        command_buffer, queue, nullptr, kernel, 1, nullptr, &num_elements,
```

Comment:
command_queue must be null according to the spec. Also applies to the rest of the clCommand* invocations.

Reply:
good spot, fixed